Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean

```diff
@@ -1,115 +1,269 @@
+  let some (type, β, inst) := args[3].app3? ``instHPow | return none
+  let arg := args[4]
+  let exp := args[5]
+  reduceAux type inst [arg] lambdas args #[inst, exp] 6 2
+      (lambdaBody? 3 >=> app6? ``HPow.hPow >=> fun (inst, a, b) =>
+        -- check for `fun a b x => a x ^ b`
+        if (a == .app (.bvar 2) (.bvar 0) && b == .bvar 1) then inst else none)
+    (fun
+      | type, [arg] => #[type, β, none, none, arg, exp]
+      | _, _ => unreachable!)
+
+
+/-- Normalize an application if the head is `+`, `*`, `-` or `/`. -/
+private def reduceHBinOp (n : Name) (args : Array Expr) (lambdas : List FVarId):
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 4 then return none else do
+  let rec instH := match n with
+    | ``HAdd.hAdd => ``instHAdd
+    | ``HMul.hMul => ``instHMul
+    | ``HSub.hSub => ``instHSub
+    | ``HDiv.hDiv => ``instHDiv
+    | _ => unreachable!
+  let some (type, inst) := args[3].app2? instH | return none
+  etaExpand args type lambdas 6 fun args lambdas _ => do
+    let lhs := args[4]
+    let rhs := args[5]
+    reduceAux type inst [lhs, rhs] lambdas args #[inst] 6 2
+      (lambdaBody? 3 >=> app6? n >=> fun (inst, l, r) =>
+        -- check for the form `fun a b x => a x + b x`
+        if (l == .app (.bvar 2) (.bvar 0) && r == .app (.bvar 1) (.bvar 0)) then inst else none)
+      (fun
+        | type, [lhs, rhs] => #[type, type, none, none, lhs, rhs]
+        | _, _ => unreachable!)
+
+/--
+Normalize an application if the head is any of `+`, `-`, `*`, `/`, `⁻¹`, `+ᵥ`, `•`, `^`.
+We assume that any lambda's at the head of the expression have already been introduced as
+free variables, and they can be specified with a list of these free variables.
+The order of the free variables is inner to outer, so `fun a b => ..` corresponds to `[b, a]`.
+
+We apply the following reductions whenever possible:
+- η-expand: increase the number of arguments in case of under application.
+  For example `HAdd.hAdd a` is turned into `fun x => a + x`.
+- Absorbing arguments: whenever a pi instance is over applied, we unfold the constant.
+  For example `(f + g) x` is turned into `f x + g x`.
+- Absorbing lambdas: lambdas in front of the constant can be moved inside.
+  For example `fun x => x + 1` is turned into `(fun x => x) + (fun x => 1)`.
+-/
+def reducePi (e : Expr) (lambdas : List FVarId) :
+    MetaM (Option (Name × List (Array (Option Expr) × List FVarId))) := do
+  if let .const n _ := e.getAppFn then
+    match n with
+    | ``Neg.neg
+    | ``Inv.inv     => Option.map (n, ·) <$> reduceUnOp   n e.getAppArgs lambdas
+    | ``HVAdd.hVAdd
+    | ``HSMul.hSMul => Option.map (n, ·) <$> reduceHActOp n e.getAppArgs lambdas
+    | ``HPow.hPow   => Option.map (n, ·) <$> reduceHPow     e.getAppArgs lambdas
+    | ``HAdd.hAdd
+    | ``HMul.hMul
+    | ``HSub.hSub
+    | ``HDiv.hDiv   => Option.map (n, ·) <$> reduceHBinOp n e.getAppArgs lambdas
```

Comment:
I think its likely we will want `inf` and `sup` here too (and to a lesser extent, all the other lattice operators). Can you either add an extension mechanism for adding more entries here, or a TODO comment about doing so?